lib/iris/cube.py

```diff
@@ -216,7 +216,30 @@ def __repr__(self):
         """Runs repr on every cube."""
         return '[%s]' % ',\n'.join([repr(cube) for cube in self])
 
+    def _check_iscube(self, obj):
```

Comment:
My take : these don't use instance properties, so they could be static methods.
In fact, they don't use private properties of Cube, so they don't really need to be in the class.
At that point they *could* go somewhere else.

For personal preference, I'd remove them from the class but keep them as private methods in cube.py, in case they might need to use 'private' cube concepts in future.

Also ... the use of `isinstance` prevents any duck typing (lookalike objects can't masquerade as Cubes), which is arguably un-Pythonic.  
We have previously used `hasattr('add_aux_coord')` for this [elsewhere](https://github.com/SciTools/iris/blob/dc468f122f8325caff5391470bf637a7aaae7adb/lib/iris/analysis/_grid_angles.py#L220)

Reply:
OK, so if we check whether the object quacks with an auxcoord instead of checking the type, is there any reason not to revert to raising an exception rather than a warning?

Reply:
Oh dear, I had somehow skimmed over that latest discussion without taking it in.
Now I see that @pelson and I are just advocating different approaches, and I honestly don't know how to choose between.

Personally though, I must day I do hate warnings in Iris.  There are still far too many, most occurrences are a pointless nuisance, and on the rare occasions when they aren't no-one is listening any more.

Reply:
No problem: I'm partly using this as a learning exercise, so exploring different solutions to the same issue is fine :+1: 

I think I prefer to have an exception on the grounds that, if my code is going to fail, it's better for it to fail sooner rather than later.  Also, having the failure at the point that the cube is included into the cubelist means that the traceback is going to point me a lot closer to where I made the mistake.   Which so far has always been
```python
cubelist.append(some_function_i_forgot_to_put_a_return_statement_in())
```

Reply:
I also agree that Iris has too many warnings!  I have actively filtered some of them out in some of my user code...

Reply:
> Also ... the use of `isinstance` prevents any duck typing (lookalike objects can't masquerade as Cubes), which is arguably un-Pythonic.
We have previously used `hasattr('add_aux_coord')` for this [elsewhere](https://github.com/SciTools/iris/blob/dc468f122f8325caff5391470bf637a7aaae7adb/lib/iris/analysis/_grid_angles.py#L220)

The ``hasattr('add_aux_coord')`` approach is, IMO, horrible. Having an ``add_aux_coord`` attribute isn't sufficient to say that a thing is cube-like. Whilst the isinstance check is somewhat irksome for duck-cubes, I'm honestly quite comfortable with getting a warning telling me that my duck doesn't inherit from the Cube. What's more, Python actually gives you the power to make your duck *claim* to inherit from the Cube, even if it doesn't (ref: https://docs.python.org/3/reference/datamodel.html#class.__instancecheck__).

@pp-mo - we should get our ducks in a row 🦆 🦆 🦆  before sending @rcomer on a wild goose chase 🤣...

@pp-mo: Can you support:

 * using ``isinstance(cube, iris.cube.Cube)`` inline within each of the necessary methods (thereby avoiding the ugly module level ``iris.cube. _check_iscube`` and associated functionality)?
 * using warnings not exceptions
 * having ``iris.cube.CubeList._warn_if_not_cube`` and ``iris.cube.CubeList._warn_if_not_iterable_of_cubes`` to do the actual checking (and thus factor out to the isinstance checks to just 2 places)? (happy enough for them to be staticmethods)

Reply:
> @pp-mo - we should get our ducks in a row duck duck duck before sending @rcomer on a wild goose chase 

I agree, but in fact I was **_strongly_** opposed to creating another warning here, because I can't see why that is good : 
What is the intended usecase where is it _useful_ to allow non-cubes, and not just raise an error ?
Also TBH @rcomer I wasn't expecting you to immediately follow my suggestion and change it all back to Exceptions !

On your other points, I really don't have strong feelings about the details.

Reply:
> Having an add_aux_coord attribute isn't sufficient to say that a thing is cube-like. 

P.S. regarding the 'hasattr/isinstance' approach, I suppose an ideal solution requires us just to identify things that are cube-like just "in respect of" their suitability for inclusion in a CubeList.  They don't have to be cube-like in all respects.
But I can't see any clear approach there, as a CubeList isn't a single-purpose thing.  The contents must support 'extract', plus whatever is needed to support merge+concatenate and and (somehow) save operations.  And other stuff, I don't have a clear list.

Reply:
> we should get our ducks in a row :duck: :duck: :duck: before sending @rcomer on a wild goose chase

nicely done! :laughing: 

As a user, I'd still rather have an exception if possible, for the reasons I gave above.

If something *does* go wrong with my cubelist, the first thing I'm going to do in an attempt to debug is print it.  So if we're looking for a minimal set of cube-like attributes, `summary` ought to be up there.

Reply:
Are we thinking about this the wrong way round?  Rather than trying to define which types should be *allowed* in a cubelist, it might be easier to focus on which types should definitely be *rejected*.  

This started because a rogue `None` in a cubelist caused problems, and I wanted a more informative error message.  So far I’m not aware of any other types that have caused issues.  So my case would be solved by simply throwing an exception `if object is None`.  We could generalise that a bit if we decide that, at minimum, the cubelist should be printable, so reject any objects that don’t have have a `summary` attribute.

The exception message needn’t say anything about how similar to a `Cube` the object is, but could just say ”object of type [whichever] does not belong in a cubelist”.

Reply:
> They don't have to be cube-like in all respects.

> Are we thinking about this the wrong way round?

I don't think so in this case. Because users can create and use their own CubeList instances, the minimum set of behaviour required for an entry into a CubeList is precisely the Cube's behaviour (and no less).

> in fact I was **_strongly_** opposed to creating another warning here

Useful to know, thank you. So my biggest concern is that we are essentially introducing a breaking change if we do this as an exception - if @rcomer has been adding ``None`` into a cube list by accident, just think of all the wild things that some of our less educated users have been doing! 😭
I guess there is a workaround though... if users **really** want to do this they can still do ``list.append(cube_list_instance, thing_that_isnt_a_cube)`` until they sort their 💩 out.

In an attempt to get consensus and prevent this conversation from being open-ended, my refined suggestion:

 * CubeList._assert_is_cube - raise a ValueError if not isinstance of cube.
 * ~~CubeList._assert_is_iterable_of_cubes~~ -> just construct a CubeList of the subset - that way you can honour iterators, and then add the constructed CubeList as necessary.
 * Update the existing call in ``CubeList.__new__`` to use ``_assert_is_cube``.

Reply:
> So my biggest concern is that we are essentially introducing a breaking change if we do this as an exception... just think of all the wild things that some of our less educated users have been doing!

I hadn't considered the possibility of cubelists being used to store random types :open_mouth: .  Semi-serious question: how far away is Iris 3?

> refined suggestion

Just to check I've understood: we make it strict so only `Cube` instances are allowed.  Because the check is restricted to one method, someone who wants to include ducks in the cubelist just needs to replace that one method?

Points 1 and 2 sound good to me.

Point 3: I think I need to wait for #3264 to be merged, and then update `__init__`!

Reply:
> @pelson  my biggest concern is that we are essentially introducing a breaking change if we do this as an exception - if @rcomer has been adding None into a cube list by accident, just think of all the wild things that some of our less educated users have been doing! sob

Not sure if it helps, but..
In my mind, if before it was _possible_ to put non-cubes into a CubeList, that was never intended behaviour, see e.g. the code covered by #3264. 
So that is a bug, and fixing a bug is not a "breaking" change.
Weaselly, but we've accepted that principle before.